scripts/abins/input/dmol3loader.py

```diff
@@ -41,245 +55,233 @@ def read_vibrational_or_phonon_data(self):
-            # We change unit of atomic displacements from atomic length units to Angstroms
-            au2ang = ATOMIC_LENGTH_2_ANGSTROM
-            float_type = FLOAT_TYPE
 
-            atoms["atom_{}".format(atom_indx)] = {"symbol": symbol, "mass": atom.mass, "sort": atom_indx,
-                                                  "coord": np.asarray(entries[1:]).astype(dtype=float_type) * au2ang}
+            # Convert coordinate units from Bohr to Angstrom
+            coord = np.asarray(coord,
+                               dtype=FLOAT_TYPE) * ATOMIC_LENGTH_2_ANGSTROM
 
-            atom_indx += 1
+            atoms[f"atom_{atom_index}"] = {"symbol": symbol, "mass": atom.mass,
+                                           "sort": atom_index, "coord": coord}
+            atom_index += 1
 
-        self.check_isotopes_substitution(atoms=atoms, masses=masses_from_file)
+        return atoms
 
-        data["atoms"] = atoms
+    @staticmethod
+    def _get_to_next_nonempty(file_obj: io.BufferedReader) -> None:
+        # Scroll to next non-empty line
+        last_line = file_obj.tell()
+        for line in file_obj:
+            if line.decode('utf8').strip():
+                file_obj.seek(last_line)
+                break
+            else:
+                last_line = file_obj.tell()
 
-    def _read_modes(self, file_obj=None, data=None):
+    @classmethod
+    def _read_modes(cls, file_obj) -> Tuple[np.ndarray, np.ndarray]:
+        """Read vibrational frequencies and atomic displacements
+
+        Modes are in their raw format: a 2-D array with modes along index 1
+        (i.e. columns, corresponding to frequencies) and atomic degrees of
+        freedom along index 0 (i.e. rows of x y z x y z x y ... for successive
+        atoms).
         """
-        Reads vibrational modes (frequencies and atomic displacements).
-        :param file_obj: file object from which we read
-        :param data: Python dictionary to which k-point data should be added
+        frequencies, displacements = cls._read_mode_block(file_obj)
+        frequency_blocks = [frequencies]
+        displacement_blocks = [displacements]
+        cls._get_to_next_nonempty(file_obj)
+
+        while not (TextParser.block_end(file_obj=file_obj, msg=["STANDARD"])
+                   or TextParser.file_end(file_obj=file_obj)):
+            frequencies, displacements = cls._read_mode_block(file_obj)
+
+            frequency_blocks.append(frequencies)
+            displacement_blocks.append(displacements)
+
+            cls._get_to_next_nonempty(file_obj)
+
+        for column in range(1, 9):
+            # Check if the last displacement block has column(s) of NaN;
+            # if so number of modes was not multiple of 9 and extra columns
+            # should be removed before assembling blocks to full array
+            if np.all(np.isnan(displacement_blocks[-1][:, column])):
+                displacement_blocks[-1] = displacement_blocks[-1][:, :column]
```

Comment:
do we not need to change the frequencies as well?

Reply:
The frequencies are read using readline.split() so always come in at the right length.

The displacements are read using a regular expression that matches 9 groups, 8 of them optional. The unused optional matches (i.e. empty columns) turn up as `None` in `match.groups()`, get converted to `NaN` by numpy.asarray() and then get cleaned up here.

Reply:
could we have a comment explaining that?

Reply:
Added one, does it look ok?